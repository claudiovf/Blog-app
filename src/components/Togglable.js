import React, { useState, useImperativeHandle } from 'react'
import PropTypes from 'prop-types'
import { Button } from 'react-bootstrap'

const Togglable = React.forwardRef((props, ref) => {
	const [visible, setVisible] = useState(false)

	const hideWhenVisible = { display: visible ? 'none' : '' }
	const showWhenVisible = { display: visible ? '' : 'none' }

	const toggleVisibility = () => {
		setVisible(!visible)
	}

	useImperativeHandle(ref, () => {
		return {
			toggleVisibility
		}
	})

	const expandContainer = props.buttonLabel !== 'view'
		?   <div style={showWhenVisible} className="visibleTogglable">
			{props.children}
			{props.title || ''} <button onClick={toggleVisibility}> {props.cancelLabel} </button>
		</div>
		:   <div style={showWhenVisible} className="visibleTogglable">
			{props.title || ''} <button onClick={toggleVisibility}> {props.cancelLabel} </button>
			{props.children}
		</div>

	return (
		<div>
			<div style={hideWhenVisible} className="togglableContent">
				{props.title} <Button variant="dark" name={props.buttonLabel} onClick={toggleVisibility}> {props.buttonLabel} </Button>
			</div>
			{expandContainer}

		</div>
	)
})

Togglable.propTypes = {
	buttonLabel: PropTypes.string.isRequired
}

Togglable.displayName = 'Togglable'

export default Togglable