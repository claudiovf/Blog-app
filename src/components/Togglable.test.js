import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import { render, fireEvent } from '@testing-library/react'
import renderer from 'react-test-renderer'
import Togglable from './Togglable'

describe('<Togglable /> ', () => {
	let component

	const tree = renderer
		.create(<Togglable buttonLabel='view'>
			<div className="testDiv"></div>
		</Togglable>)
		.toJSON()

	expect(tree).toMatchSnapshot()

	beforeEach(() => {
		component = render(
			<Togglable buttonLabel='view'>
				<div className='testDiv' />
			</Togglable>
		)
	})

	test('renders its children', () => {
		expect(component.container.querySelector('.testDiv')
		).toBeDefined()
	})
	test('when blog is visible', () => {
		const button = component.getByText('view')
		fireEvent.click(button)

		const div = component.container.querySelector('.testDiv')

		expect(div).not.toHaveStyle('display: none')

	})
})