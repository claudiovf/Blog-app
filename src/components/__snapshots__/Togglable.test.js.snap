// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <div
    className="togglableContent"
    style={
      Object {
        "display": "",
      }
    }
  >
     
    <button
      onClick={[Function]}
    >
       
      view
       
    </button>
  </div>
  <div
    className="visibleTogglable"
    style={
      Object {
        "display": "none",
      }
    }
  >
    
     
    <button
      onClick={[Function]}
    >
       
       
    </button>
    <div
      className="testDiv"
    />
  </div>
</div>
`;
